plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java

```diff
@@ -0,0 +1,680 @@
+        Map<String, String> modelConfigMaps,
+        int maxPredictionTask,
+        String tag,
+        String description,
+        boolean ignoreMissing,
+        String functionName,
+        boolean fullResponsePath,
+        boolean ignoreFailure,
+        boolean override,
+        String modelInput,
+        Client client,
+        NamedXContentRegistry xContentRegistry
+    ) {
+        super(tag, description, ignoreFailure);
+        this.inferenceProcessorAttributes = new InferenceProcessorAttributes(
+            modelId,
+            inputMaps,
+            outputMaps,
+            modelConfigMaps,
+            maxPredictionTask
+        );
+        this.ignoreMissing = ignoreMissing;
+        this.functionName = functionName;
+        this.fullResponsePath = fullResponsePath;
+        this.ignoreFailure = ignoreFailure;
+        this.override = override;
+        this.modelInput = modelInput;
+        this.client = client;
+        this.xContentRegistry = xContentRegistry;
+    }
+
+    @Override
+    public SearchResponse processResponse(SearchRequest request, SearchResponse response) throws Exception {
+        throw new RuntimeException("ML inference search response processor make asynchronous calls and does not call processRequest");
+    }
+
+    /**
+     * Processes the search response asynchronously by rewriting the documents with the inference results.
+     *
+     * @param request         the search request
+     * @param response        the search response
+     * @param responseContext the pipeline processing context
+     * @param responseListener the listener to be notified when the response is processed
+     */
+    @Override
+    public void processResponseAsync(
+        SearchRequest request,
+        SearchResponse response,
+        PipelineProcessingContext responseContext,
+        ActionListener<SearchResponse> responseListener
+    ) {
+        try {
+            SearchHit[] hits = response.getHits().getHits();
+            // skip processing when there is no hit
+            if (hits.length == 0) {
+                responseListener.onResponse(response);
+                return;
+            }
+            rewriteResponseDocuments(response, hits, responseListener);
+        } catch (Exception e) {
+            if (ignoreFailure) {
+                responseListener.onResponse(response);
+            } else {
+                responseListener.onFailure(e);
+            }
+        }
+    }
+
+    /**
+     * Rewrite the documents in the search response with the inference results.
+     *
+     * @param response        the search response
+     * @param hits            the search hits
+     * @param responseListener the listener to be notified when the response is processed
+     * @throws IOException if an I/O error occurs during the rewriting process
+     */
+    private void rewriteResponseDocuments(SearchResponse response, SearchHit[] hits, ActionListener<SearchResponse> responseListener)
```

Comment:
Why do we pass both response and hits? I think response already includes hits.

Reply:
remove the response here

Reply:
So remove hits? The response has hits.

Reply:
removed!